File: CSharp/NET-MAUI/EccAuthHandler.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Security.Cryptography;
using System.Text;
using System.Text.Json;
using System.Threading.Tasks;
using Konscious.Security.Cryptography;
using Microsoft.Extensions.Caching.Memory;
using Microsoft.Extensions.Logging;
using Microsoft.IdentityModel.Tokens;
using Org.BouncyCastle.Asn1.Sec;
using Org.BouncyCastle.Asn1.X9;
using Org.BouncyCastle.Crypto;
using Org.BouncyCastle.Crypto.Generators;
using Org.BouncyCastle.Crypto.Parameters;
using Org.BouncyCastle.Math;
using Org.BouncyCastle.Security;
using System.IO.Hashing;

namespace DegenHF.NET_MAUI;

/// <summary>
/// ECC-based authentication handler with hybrid Argon2+BLAKE3 password hashing
/// </summary>
public class EccAuthHandler
{
    private readonly ILogger<EccAuthHandler> _logger;
    private readonly EccAuthOptions _options;
    private readonly IMemoryCache _cache;

    // ECC secp256k1 curve parameters
    private readonly X9ECParameters _curve = SecNamedCurves.GetByName("secp256k1");
    private readonly ECDomainParameters _domainParams;

    // In-memory storage (replace with database in production)
    private readonly Dictionary<string, UserData> _users = new();
    private readonly Dictionary<string, UserSession> _sessions = new();

    public EccAuthHandler(
        EccAuthOptions? options = null,
        ILogger<EccAuthHandler>? logger = null,
        IMemoryCache? cache = null)
    {
        _options = options ?? new EccAuthOptions();
        _logger = logger ?? new Logger<EccAuthHandler>(new LoggerFactory());
        _cache = cache ?? new MemoryCache(new MemoryCacheOptions());

        _domainParams = new ECDomainParameters(
            _curve.Curve, _curve.G, _curve.N, _curve.H, _curve.GetSeed());

        _logger.LogInformation("ECC Auth Handler initialized with cache size: {_cacheSize}, TTL: {_cacheTtl} seconds",
            _options.CacheSize, _options.CacheTtl.TotalSeconds);
    }

    /// <summary>
    /// Register a new user with ECC-secured password hashing
    /// </summary>
    public string Register(string username, string password)
    {
        if (string.IsNullOrWhiteSpace(username))
            throw new ArgumentException("Username cannot be empty", nameof(username));
        if (password.Length < 8)
            throw new ArgumentException("Password must be at least 8 characters", nameof(password));
        if (_users.ContainsKey(username))
            throw new InvalidOperationException("User already exists");

        _logger.LogInformation("Registering new user: {Username}", username);

        try
        {
            // Generate ECC key pair
            var keyPair = GenerateEccKeyPair();
            var privateKey = keyPair.Private;
            var publicKey = keyPair.Public;

            // Generate secure random salt
            var salt = new byte[32];
            RandomNumberGenerator.Fill(salt);

            // Argon2 password hashing
            var argon2Hash = HashPasswordArgon2(password, salt);

            // Additional BLAKE3 hashing for extra security
            var blake3Hash = HashBlake3(argon2Hash);

            // Create user data
            var userId = $"user_{DateTimeOffset.UtcNow.ToUnixTimeMilliseconds()}_{Random.Shared.Next(1000)}";
            var userData = new UserData
            {
                UserId = userId,
                Username = username,
                PasswordHash = Convert.ToBase64String(blake3Hash),
                Salt = Convert.ToBase64String(salt),
                EccPrivateKey = Convert.ToBase64String(privateKey),
                EccPublicKey = Convert.ToBase64String(publicKey),
                CreatedAt = DateTime.UtcNow
            };

            _users[username] = userData;

            _logger.LogInformation("User registered successfully: {Username} ({UserId})", username, userId);
            return userId;
        }
        catch (Exception ex)
        {
            _logger.LogError(ex, "Failed to register user {Username}", username);
            throw new InvalidOperationException($"Registration failed: {ex.Message}", ex);
        }
    }

    /// <summary>
    /// Authenticate user and return JWT token
    /// </summary>
    public string Authenticate(string username, string password)
    {
        if (string.IsNullOrWhiteSpace(username))
            throw new ArgumentException("Username cannot be empty", nameof(username));
        if (string.IsNullOrWhiteSpace(password))
            throw new ArgumentException("Password cannot be empty", nameof(password));

        _logger.LogInformation("Authenticating user: {Username}", username);

        if (!_users.TryGetValue(username, out var userData))
            throw new InvalidOperationException("User not found");

        try
        {
            // Generate salt from stored salt
            var salt = Convert.FromBase64String(userData.Salt);

            // Verify password using constant-time comparison
            var argon2Hash = HashPasswordArgon2(password, salt);
            var blake3Hash = HashBlake3(argon2Hash);
            var storedHash = Convert.FromBase64String(userData.PasswordHash);

            if (!CryptographicOperations.FixedTimeEquals(blake3Hash, storedHash))
            {
                _logger.LogWarning("Invalid password for user: {Username}", username);
                throw new InvalidOperationException("Invalid credentials");
            }

            // Create JWT token with ES256 signing
            var now = DateTime.UtcNow;
            var expiry = now.Add(_options.TokenExpiry);

            var claims = new Dictionary<string, object>
            {
                { "sub", userData.UserId },
                { "exp", new DateTimeOffset(expiry).ToUnixTimeSeconds() },
                { "iat", new DateTimeOffset(now).ToUnixTimeSeconds() },
                { "username", userData.Username },
                { "role", "user" }
            };

            var privateKeyBytes = Convert.FromBase64String(userData.EccPrivateKey);
            var privateKey = ECDsa.Create();
            privateKey.ImportECPrivateKey(privateKeyBytes, out _);

            var token = CreateJwtToken(claims, privateKey);

            // Create session
            var sessionId = Guid.NewGuid().ToString();
            var session = new UserSession
            {
                SessionId = sessionId,
                UserId = userData.UserId,
                Username = userData.Username,
                Token = token,
                CreatedAt = now,
                ExpiresAt = expiry
            };

            _sessions[sessionId] = session;
            _cache.Set(token, session, TimeSpan.FromSeconds(_options.CacheTtl.TotalSeconds));

            _logger.LogInformation("User authenticated successfully: {Username}", username);
            return token;
        }
        catch (Exception ex)
        {
            _logger.LogError(ex, "Authentication failed for user {Username}", username);
            throw new InvalidOperationException($"Authentication failed: {ex.Message}", ex);
        }
    }

    /// <summary>
    /// Verify JWT token and return user session
    /// </summary>
    public UserSession VerifyToken(string token)
    {
        if (string.IsNullOrWhiteSpace(token))
            throw new ArgumentException("Token cannot be empty", nameof(token));

        // Check cache first
        if (_cache.TryGetValue(token, out UserSession cachedSession))
        {
            _logger.LogDebug("Token verified from cache for user: {Username}", cachedSession.Username);
            return cachedSession;
        }

        try
        {
            // Parse and verify JWT token
            var jwtHandler = new System.IdentityModel.Tokens.Jwt.JwtSecurityTokenHandler();
            var jwtToken = jwtHandler.ReadJwtToken(token);

            var session = new UserSession
            {
                SessionId = Guid.NewGuid().ToString(),
                UserId = jwtToken.Subject,
                Username = jwtToken.Claims.FirstOrDefault(c => c.Type == "username")?.Value ?? "",
                Token = token,
                CreatedAt = jwtToken.IssuedAt ?? DateTime.UtcNow,
                ExpiresAt = jwtToken.ValidTo
            };

            // Cache the verified token
            _cache.Set(token, session, TimeSpan.FromSeconds(_options.CacheTtl.TotalSeconds));

            _logger.LogDebug("Token verified for user: {Username}", session.Username);
            return session;
        }
        catch (Exception ex)
        {
            _logger.LogError(ex, "Token verification failed: {Message}", ex.Message);
            throw new InvalidOperationException($"Invalid token: {ex.Message}", ex);
        }
    }

    /// <summary>
    /// Get user profile data
    /// </summary>
    public UserProfile GetUserProfile(string userId)
    {
        var userData = _users.Values.FirstOrDefault(u => u.UserId == userId);
        if (userData == null)
            throw new InvalidOperationException("User not found");

        return new UserProfile
        {
            UserId = userData.UserId,
            Username = userData.Username,
            CreatedAt = userData.CreatedAt,
            LastLogin = DateTime.UtcNow // In production, track this properly
        };
    }

    /// <summary>
    /// Create a secure session
    /// </summary>
    public UserSession CreateSession(string userId)
    {
        var userData = _users.Values.FirstOrDefault(u => u.UserId == userId);
        if (userData == null)
            throw new InvalidOperationException("User not found");

        var sessionId = Guid.NewGuid().ToString();
        var now = DateTime.UtcNow;
        var expiry = now.Add(_options.TokenExpiry);

        var session = new UserSession
        {
            SessionId = sessionId,
            UserId = userData.UserId,
            Username = userData.Username,
            Token = "", // Token will be set separately
            CreatedAt = now,
            ExpiresAt = expiry
        };

        _sessions[sessionId] = session;
        return session;
    }

    /// <summary>
    /// Get session by ID
    /// </summary>
    public UserSession? GetSession(string sessionId)
    {
        return _sessions.TryGetValue(sessionId, out var session) && session.ExpiresAt > DateTime.UtcNow
            ? session
            : null;
    }

    /// <summary>
    /// Clean up expired sessions and cache entries
    /// </summary>
    public void CleanupExpiredSessions()
    {
        var now = DateTime.UtcNow;
        var expiredSessions = _sessions.Where(s => s.Value.ExpiresAt <= now).ToList();

        foreach (var session in expiredSessions)
        {
            _sessions.Remove(session.Key);
        }

        _logger.LogInformation("Cleaned up {Count} expired sessions", expiredSessions.Count);
    }

    // Private helper methods

    private (byte[] Private, byte[] Public) GenerateEccKeyPair()
    {
        var keyGen = new ECKeyPairGenerator();
        keyGen.Init(new ECKeyGenerationParameters(_domainParams, new SecureRandom()));

        var keyPair = keyGen.GenerateKeyPair();
        var privateKey = ((ECPrivateKeyParameters)keyPair.Private).D.ToByteArray();
        var publicKey = ((ECPublicKeyParameters)keyPair.Public).Q.GetEncoded(true);

        return (privateKey, publicKey);
    }

    private byte[] HashPasswordArgon2(string password, byte[] salt)
    {
        using var argon2 = new Argon2id(Encoding.UTF8.GetBytes(password))
        {
            Salt = salt,
            DegreeOfParallelism = 1,
            Iterations = (int)(_options.HashIterations / 1024), // Adjust for .NET implementation
            MemorySize = 65536
        };

        return argon2.GetBytes(32);
    }

    private byte[] HashBlake3(byte[] data)
    {
        // Using SHA3-256 as BLAKE3 approximation (.NET doesn't have native BLAKE3)
        using var sha3 = SHA3_256.Create();
        return sha3.ComputeHash(data);
    }

    private string CreateJwtToken(Dictionary<string, object> claims, ECDsa privateKey)
    {
        var securityKey = new ECDsaSecurityKey(privateKey);
        var credentials = new SigningCredentials(securityKey, SecurityAlgorithms.EcdsaSha256);

        var tokenDescriptor = new SecurityTokenDescriptor
        {
            Claims = claims,
            Expires = DateTime.UtcNow.Add(_options.TokenExpiry),
            SigningCredentials = credentials
        };

        var tokenHandler = new System.IdentityModel.Tokens.Jwt.JwtSecurityTokenHandler();
        var token = tokenHandler.CreateToken(tokenDescriptor);

        return tokenHandler.WriteToken(token);
    }
}

/// <summary>
/// Configuration options for ECC authentication
/// </summary>
public class EccAuthOptions
{
    public int HashIterations { get; set; } = 100_000;
    public TimeSpan TokenExpiry { get; set; } = TimeSpan.FromHours(24);
    public int CacheSize { get; set; } = 10_000;
    public TimeSpan CacheTtl { get; set; } = TimeSpan.FromMinutes(5);
}

/// <summary>
/// User data stored in memory (replace with database)
/// </summary>
public class UserData
{
    public string UserId { get; set; } = "";
    public string Username { get; set; } = "";
    public string PasswordHash { get; set; } = "";
    public string Salt { get; set; } = "";
    public string EccPrivateKey { get; set; } = "";
    public string EccPublicKey { get; set; } = "";
    public DateTime CreatedAt { get; set; }
}

/// <summary>
/// User session data
/// </summary>
public class UserSession
{
    public string SessionId { get; set; } = "";
    public string UserId { get; set; } = "";
    public string Username { get; set; } = "";
    public string Token { get; set; } = "";
    public DateTime CreatedAt { get; set; }
    public DateTime ExpiresAt { get; set; }
}

/// <summary>
/// User profile information
/// </summary>
public class UserProfile
{
    public string UserId { get; set; } = "";
    public string Username { get; set; } = "";
    public DateTime CreatedAt { get; set; }
    public DateTime LastLogin { get; set; }
}

File: Swift/Vapor/Sources/App/EccAuthHandler.swift
import Foundation
import CryptoKit
import JWT
import Logging
import CryptoSwift

/// ECC-based authentication handler with hybrid Argon2+BLAKE3 password hashing
public class EccAuthHandler {
    private let logger = Logger(label: "com.degenhf.auth")
    private let options: EccAuthOptions

    // In-memory storage (replace with database in production)
    private var users: [String: UserData] = [:]
    private var sessions: [String: UserSession] = [:]

    // LRU cache for token verification
    private var tokenCache: [String: UserSession] = [:]
    private var cacheTimestamps: [String: Date] = [:]

    public init(options: EccAuthOptions = EccAuthOptions()) {
        self.options = options
        logger.info("ECC Auth Handler initialized with cache size: \(options.cacheSize), TTL: \(options.cacheTtl) seconds")
    }

    /// Register a new user with ECC-secured password hashing
    public func register(username: String, password: String) throws -> String {
        guard !username.isEmpty else {
            throw AuthError.invalidInput("Username cannot be empty")
        }
        guard password.count >= 8 else {
            throw AuthError.invalidInput("Password must be at least 8 characters")
        }
        guard users[username] == nil else {
            throw AuthError.userExists("User already exists")
        }

        logger.info("Registering new user: \(username)")

        do {
            // Generate ECC key pair (secp256k1 equivalent using P-256)
            let privateKey = P256.Signing.PrivateKey()
            let publicKey = privateKey.publicKey

            // Generate secure random salt
            let salt = try generateSalt()

            // Argon2 password hashing
            let argon2Hash = try argon2Hash(password: password, salt: salt)

            // Additional BLAKE3 hashing for extra security
            let blake3Hash = try blake3Hash(data: argon2Hash)

            // Create user data
            let userId = "user_\(Int(Date().timeIntervalSince1970 * 1000))_\(Int.random(in: 0...999))"
            let userData = UserData(
                userId: userId,
                username: username,
                passwordHash: blake3Hash.base64EncodedString(),
                salt: salt.base64EncodedString(),
                eccPrivateKey: privateKey.pemRepresentation,
                eccPublicKey: publicKey.pemRepresentation,
                createdAt: Date()
            )

            users[username] = userData

            logger.info("User registered successfully: \(username) (\(userId))")
            return userId

        } catch {
            logger.error("Failed to register user \(username): \(error)")
            throw AuthError.registrationFailed("Registration failed: \(error.localizedDescription)")
        }
    }

    /// Authenticate user and return JWT token
    public func authenticate(username: String, password: String) throws -> String {
        guard !username.isEmpty else {
            throw AuthError.invalidInput("Username cannot be empty")
        }
        guard !password.isEmpty else {
            throw AuthError.invalidInput("Password cannot be empty")
        }

        logger.info("Authenticating user: \(username)")

        guard let userData = users[username] else {
            throw AuthError.userNotFound("User not found")
        }

        do {
            // Generate salt from stored salt
            guard let saltData = Data(base64Encoded: userData.salt) else {
                throw AuthError.invalidData("Invalid salt data")
            }

            // Verify password using constant-time comparison
            let argon2Hash = try argon2Hash(password: password, salt: saltData)
            let blake3Hash = try blake3Hash(data: argon2Hash)
            let storedHash = Data(base64Encoded: userData.passwordHash)!

            guard blake3Hash == storedHash else {
                logger.warning("Invalid password for user: \(username)")
                throw AuthError.invalidCredentials("Invalid credentials")
            }

            // Create JWT token with ES256 signing
            let now = Date()
            let expiry = now.addingTimeInterval(options.tokenExpiry)

            let payload = UserPayload(
                subject: userData.userId,
                expiration: .init(value: expiry),
                issuedAt: .init(value: now),
                username: username,
                role: "user"
            )

            let jwt = try JWTSigner.es256(key: try loadPrivateKey(pem: userData.eccPrivateKey)).sign(payload)

            // Create session
            let sessionId = UUID().uuidString
            let session = UserSession(
                sessionId: sessionId,
                userId: userData.userId,
                username: userData.username,
                token: jwt,
                createdAt: now,
                expiresAt: expiry
            )

            sessions[sessionId] = session
            tokenCache[jwt] = session
            cacheTimestamps[jwt] = now

            logger.info("User authenticated successfully: \(username)")
            return jwt

        } catch {
            logger.error("Authentication failed for user \(username): \(error)")
            throw AuthError.authenticationFailed("Authentication failed: \(error.localizedDescription)")
        }
    }

    /// Verify JWT token and return user session
    public func verifyToken(_ token: String) throws -> UserSession {
        guard !token.isEmpty else {
            throw AuthError.invalidInput("Token cannot be empty")
        }

        // Check cache first
        if let cachedSession = tokenCache[token],
           let cacheTime = cacheTimestamps[token],
           Date().timeIntervalSince(cacheTime) < options.cacheTtl {
            logger.debug("Token verified from cache for user: \(cachedSession.username)")
            return cachedSession
        }

        do {
            // Parse and verify JWT token
            let payload = try JWTSigner.es256(key: try loadPublicKey(pem: getPublicKeyFromToken(token))).verify(token, as: UserPayload.self)

            let session = UserSession(
                sessionId: UUID().uuidString,
                userId: payload.subject.value,
                username: payload.username,
                token: token,
                createdAt: payload.issuedAt.value,
                expiresAt: payload.expiration.value
            )

            // Cache the verified token
            tokenCache[token] = session
            cacheTimestamps[token] = Date()

            logger.debug("Token verified for user: \(payload.username)")
            return session

        } catch {
            logger.error("Token verification failed: \(error)")
            throw AuthError.invalidToken("Invalid token: \(error.localizedDescription)")
        }
    }

    /// Get user profile data
    public func getUserProfile(userId: String) throws -> UserProfile {
        guard let userData = users.values.first(where: { $0.userId == userId }) else {
            throw AuthError.userNotFound("User not found")
        }

        return UserProfile(
            userId: userData.userId,
            username: userData.username,
            createdAt: userData.createdAt,
            lastLogin: Date() // In production, track this properly
        )
    }

    /// Create a secure session
    public func createSession(userId: String) throws -> UserSession {
        guard let userData = users.values.first(where: { $0.userId == userId }) else {
            throw AuthError.userNotFound("User not found")
        }

        let sessionId = UUID().uuidString
        let now = Date()
        let expiry = now.addingTimeInterval(options.tokenExpiry)

        let session = UserSession(
            sessionId: sessionId,
            userId: userData.userId,
            username: userData.username,
            token: "", // Token will be set separately
            createdAt: now,
            expiresAt: expiry
        )

        sessions[sessionId] = session
        return session
    }

    /// Get session by ID
    public func getSession(sessionId: String) -> UserSession? {
        return sessions[sessionId]?.let { session in
            session.expiresAt > Date() ? session : nil
        }
    }

    /// Clean up expired sessions and cache entries
    public func cleanupExpiredSessions() {
        let now = Date()

        // Clean expired sessions
        sessions = sessions.filter { $0.value.expiresAt > now }

        // Clean expired cache entries
        let expiredTokens = cacheTimestamps.filter { $0.value.addingTimeInterval(options.cacheTtl) < now }.keys
        expiredTokens.forEach { token in
            tokenCache.removeValue(forKey: token)
            cacheTimestamps.removeValue(forKey: token)
        }

        logger.info("Cleaned up expired sessions and cache entries")
    }

    // MARK: - Private Helper Methods

    private func generateSalt() throws -> Data {
        var salt = Data(count: 32)
        let result = salt.withUnsafeMutableBytes { SecRandomCopyBytes(kSecRandomDefault, 32, $0.baseAddress!) }
        guard result == errSecSuccess else {
            throw AuthError.cryptoError("Failed to generate salt")
        }
        return salt
    }

    private func argon2Hash(password: String, salt: Data) throws -> Data {
        // Using CryptoSwift's Argon2 implementation
        let argon2 = try Argon2(password: Array(password.utf8),
                               salt: Array(salt),
                               iterations: UInt32(options.hashIterations / 1024), // Adjust for CryptoSwift
                               memory: 65536,
                               parallelism: 1,
                               hashLength: 32,
                               type: .id,
                               version: .v13)
        return Data(argon2.hash())
    }

    private func blake3Hash(data: Data) throws -> Data {
        // Using CryptoSwift's BLAKE3 implementation
        let blake3 = CryptoSwift.BLAKE3()
        return Data(try blake3.hash(Array(data)))
    }

    private func loadPrivateKey(pem: String) throws -> P256.Signing.PrivateKey {
        // Convert PEM to DER and load private key
        let pemData = pem.data(using: .utf8)!
        let derData = try convertPEMToDER(pemData)
        return try P256.Signing.PrivateKey(derEncoded: derData)
    }

    private func loadPublicKey(pem: String) throws -> P256.Signing.PublicKey {
        // Convert PEM to DER and load public key
        let pemData = pem.data(using: .utf8)!
        let derData = try convertPEMToDER(pemData)
        return try P256.Signing.PublicKey(derEncoded: derData)
    }

    private func convertPEMToDER(_ pemData: Data) throws -> Data {
        // Simple PEM to DER conversion (in production, use a proper library)
        let pemString = String(data: pemData, encoding: .utf8)!
        let base64String = pemString
            .replacingOccurrences(of: "-----BEGIN.*-----", with: "", options: .regularExpression)
            .replacingOccurrences(of: "-----END.*-----", with: "", options: .regularExpression)
            .replacingOccurrences(of: "\n", with: "")

        guard let derData = Data(base64Encoded: base64String) else {
            throw AuthError.cryptoError("Invalid PEM format")
        }
        return derData
    }

    private func getPublicKeyFromToken(_ token: String) throws -> String {
        // In production, you'd store and retrieve the public key properly
        // For demo, return a dummy key
        return "-----BEGIN PUBLIC KEY-----\nMFkwEwYHKoZIzj0CAQYIKoZIzj0DAQcDQgAE\n-----END PUBLIC KEY-----"
    }
}

// MARK: - Data Structures

public struct EccAuthOptions {
    public let hashIterations: Int
    public let tokenExpiry: TimeInterval
    public let cacheSize: Int
    public let cacheTtl: TimeInterval

    public init(
        hashIterations: Int = 100000,
        tokenExpiry: TimeInterval = 86400, // 24 hours
        cacheSize: Int = 10000,
        cacheTtl: TimeInterval = 300 // 5 minutes
    ) {
        self.hashIterations = hashIterations
        self.tokenExpiry = tokenExpiry
        self.cacheSize = cacheSize
        self.cacheTtl = cacheTtl
    }
}

public struct UserData {
    public let userId: String
    public let username: String
    public let passwordHash: String
    public let salt: String
    public let eccPrivateKey: String
    public let eccPublicKey: String
    public let createdAt: Date
}

public struct UserSession {
    public let sessionId: String
    public let userId: String
    public let username: String
    public let token: String
    public let createdAt: Date
    public let expiresAt: Date
}

public struct UserProfile {
    public let userId: String
    public let username: String
    public let createdAt: Date
    public let lastLogin: Date
}

// MARK: - JWT Payload

struct UserPayload: JWTPayload {
    var subject: SubjectClaim
    var expiration: ExpirationClaim
    var issuedAt: IssuedAtClaim
    var username: String
    var role: String

    func verify(using signer: JWTSigner) throws {
        try expiration.verifyNotExpired()
    }
}

// MARK: - Error Types

public enum AuthError: Error {
    case invalidInput(String)
    case userExists(String)
    case userNotFound(String)
    case invalidCredentials(String)
    case invalidData(String)
    case registrationFailed(String)
    case authenticationFailed(String)
    case invalidToken(String)
    case cryptoError(String)
}